### DISCUSSION TITLE
nnn opening file in firefox even though xdg-mime is set correctly

### DISCUSSION THREAD

**Discussion by schaschko** [jarun/nnn#1641] [Q&A] [answered]
I have a file with filetype `application/json`. I checked this with `xdg-mime query filetype $file`. When using `xdg-open $file` it opens correctly in `application/json=vim.desktop`, that I set in `mimeapps.list`.

In `nnn`, this file will then open with firefox when using `l`, the open command. When using `e` it opens in vim as expected. I researched `nnn` is calling `xdg-open`, so the file should open with `vim`, but not `firefox` as it does now.

I could debug to the open where I use "open with" in `nnn` (press key `o`) and open with `xdg-open`, then it asks wether it should open with gui or cli mode. When opened in cli, then vim is started which is good, if it is gui, then it is firefox, not good.

It seems that `nnn` is not using `xdg-open` when using `l`. I want `nnn` to always open vim when opening a file with the key `l`. How to achieve that?

**Comment by jarun** [comment]
Can you please compile `nnn` in debug mode and share the logs?

Also, there are some troubleshooting steps here: https://github.com/jarun/nnn/wiki/Troubleshooting#wrong-file-opener

**Comment by schaschko** [comment]
I could solve the problem by running `nnn -c` instead of just `nnn`, but I think it should also work without `-c`. Calling `xdg-open` on the file directly (e.g from my shell) always opens with correct application. Is `nnn` really always calling `xdg-open` on pressing `l`?

Nothing going on in the log file I think (running `./nnn` after building with DEBUG (without `-c`)):

```
ln 8642: VERSION=4.8
ln 8431: g_tmpfpath=/tmp
ln 8432: tmpfplen=5
ln 8675: home=/home/asdf
ln 8379: cfgpath=/home/asdf/.config
ln 8385: cfgpath=/home/asdf/.config/nnn
ln 8413: selpath=/home/asdf/.config/nnn/.selection
ln 8683: opener=xdg-open
ln 8793: getenv(envs[ENV_VISUAL])=(null)
ln 8794: getenv(envs[ENV_EDITOR])=vim
ln 8795: editor=vim
ln 8799: pager=less -R
ln 8803: shell=/usr/bin/fish
ln 8805: getenv("PWD")=/home/asdf/tmp/nnn
ln 2226: COLORS=256
ln 2227: COLOR_PAIRS=65536
ln 5688: __func__=dentfill
ln 5955: ts2.tv_nsec - ts1.tv_nsec=107988
ln 6545: __func__=redraw
ln 6560: path=/home/asdf/tmp/nnn
ln 7017: newpath=/home/asdf/tmp/nnn/tsconfig.json
ln 7035: sb.st_mode=33152
ln 2463: pid=1709983
ln 6545: __func__=redraw
ln 6560: path=/home/asdf/tmp/nnn
```

**Reply by jarun** [reply]
Without `-c` the opener is indeed `xdg-open`.

Can you share the log with `-c` to compare?

**Reply by jarun** [reply]
Ahh OK. Maybe vim is opening in the background. `-c` brings it to the foreground.

**Reply by schaschko** [reply]
Pressing `l` without `-c` opens firefox unfortunately.

Log with `-c`, seems the same as without:

```
ln 8642: VERSION=4.8
ln 8431: g_tmpfpath=/tmp
ln 8432: tmpfplen=5
ln 8675: home=/home/asdf
ln 8379: cfgpath=/home/asdf/.config
ln 8385: cfgpath=/home/asdf/.config/nnn
ln 8413: selpath=/home/asdf/.config/nnn/.selection
ln 8683: opener=xdg-open
ln 8793: getenv(envs[ENV_VISUAL])=(null)
ln 8794: getenv(envs[ENV_EDITOR])=vim
ln 8795: editor=vim
ln 8799: pager=less -R
ln 8803: shell=/usr/bin/fish
ln 8805: getenv("PWD")=/home/asdf/tmp/nnn
ln 2226: COLORS=256
ln 2227: COLOR_PAIRS=65536
ln 5688: __func__=dentfill
ln 5955: ts2.tv_nsec - ts1.tv_nsec=112608
ln 6545: __func__=redraw
ln 6560: path=/home/asdf/tmp/nnn
ln 7017: newpath=/home/asdf/tmp/nnn/tsconfig.json
ln 7035: sb.st_mode=33152
ln 2389: status=0
ln 2463: pid=1755597
ln 6545: __func__=redraw
ln 6560: path=/home/asdf/tmp/nnn
```

**Reply by jarun** [reply, marked as answer]
Yes, `-c` enforces the cli mode. Without `-c` a gui app is chosen by xdg-open it seems. xdg-open is the opener with or without `-c`.
The vim window doesn't show without `-c` because nnn tries to open it as a gui app.

**Reply by schaschko** [reply]
From the [arch wiki](https://wiki.archlinux.org/title/default_applications#xdg-open):

> Because of the complexity of the [xdg-utils](https://wiki.archlinux.org/title/Xdg-utils) version of [xdg-open](https://wiki.archlinux.org/title/Xdg-open), it can be difficult to debug when the wrong default application is being opened

Problem seems to be with `xdg-open`, not `nnn`.

Thank you
